src/Features/Core/Diagnostics/EngineV1/DiagnosticAnalyzerDriver.cs

```diff
@@ -445,27 +445,55 @@ public bool IsAnalyzerSuppressed(DiagnosticAnalyzer analyzer)
 
-                // Get all the analyzer actions, including the per-compilation actions.
-                var analyzerExecutor = GetAnalyzerExecutor(analyzer, compilation, localDiagnostics.Add);
-                var analyzerActions = await this.GetAnalyzerActionsAsync(analyzer, analyzerExecutor).ConfigureAwait(false);
-                var hasDependentCompilationEndAction = await AnalyzerManager.Instance.GetAnalyzerHasDependentCompilationEndAsync(analyzer, analyzerExecutor).ConfigureAwait(false);
-
-                if (hasDependentCompilationEndAction && forceAnalyzeAllDocuments != null)
+                try
                 {
-                    // Analyzer registered a compilation end action and at least one other analyzer action during its compilation start action.
-                    // We need to ensure that we have force analyzed all documents in this project for this analyzer before executing the end actions.
-                    forceAnalyzeAllDocuments(_project, analyzer, _cancellationToken);
-                }
+                    // Get all the analyzer actions, including the per-compilation actions.
+                    var analyzerExecutor = GetAnalyzerExecutor(analyzer, compilation, localDiagnostics.Add);
+                    var analyzerActions = await this.GetAnalyzerActionsAsync(analyzer, analyzerExecutor).ConfigureAwait(false);
+                    var hasDependentCompilationEndAction = await AnalyzerManager.Instance.GetAnalyzerHasDependentCompilationEndAsync(analyzer, analyzerExecutor).ConfigureAwait(false);
+
+                    if (hasDependentCompilationEndAction && forceAnalyzeAllDocuments != null)
+                    {
+                        // Analyzer registered a compilation end action and at least one other analyzer action during its compilation start action.
+                        // We need to ensure that we have force analyzed all documents in this project for this analyzer before executing the end actions.
+                        // Doing so on the original analyzer instance might cause duplicate callbacks into the analyzer.
+                        // So we create a new instance of this analyzer and compute compilation diagnostics on that instance.
+
+                        try
+                        {
+                            analyzer = Activator.CreateInstance(analyzer.GetType()) as DiagnosticAnalyzer;
+                        }
+                        catch
+                        {
+                            // Unable to created a new analyzer instance, bail out on reporting diagnostics.
+                            return;
+                        }
+
+                        analyzerExecutor = GetAnalyzerExecutor(analyzer, compilation, localDiagnostics.Add);
+                        analyzerActions = await this.GetAnalyzerActionsAsync(analyzer, analyzerExecutor).ConfigureAwait(false);
+                        analyzerToDispose = analyzer;
```

Comment:
We have specified that Initialize will be called no more than once per analyzer instance. It's hard to predict what might break as a consequence of violating that. One might imagine an analyzer that acquires some resource in Initialize and frees it in a finalizer. Calling Initialize more than once for an instance of such an analyzer could leak the resource.